sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -559,6 +557,19 @@ case class BinaryPredicate(override val prettyName: String, left: Expression, ri
   }
 }
 
+trait CollationCheck {
+  self: BinaryExpression =>
```

Comment:
I think that you should implement this against `StringPredicate` instead of `BinaryExpression`. If it is just `BinaryExpression` you don't have any guarantee that `asInstanceOf[StringType]` will work.

Reply:
Btw, good job adding this trait. We do need this!